### PR TITLE
Changes to persist TSA/B state across reloads

### PR DESCRIPTION
**Pull request by tjchadaga**
#### Why I did it
Changes to save TSA state in configDB and persist TSA/B across reboots

#### How I did it
* Added cfgdb table to hold TSA/B state
* Added handling in bgpcfg to subscribe to TSA state change
* Added test code and data files

#### How to verify it
* Execute TSA, config save, reboot or config reload.  After system comes up, check TSC to ensure the system is in Maintenance
 
#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [x] 201911
- [ ] 202006
- [x] 202012
- [x] 202106
- [ ] 202111
- [x] 202205

### CODE REVIEW

**Review by arlakshm — COMMENTED** [review]
as comments

File: dockers/docker-fpm-frr/base_image_files/TS [hunk 1]
```diff
@@ -20,10 +26,19 @@ if [[ ($NUM_ASIC -gt 1) ]]; then
         if [ $sub_role == 'FrontEnd' ]
         then
             echo -e "BGP"$asic" : \c"
-            docker exec -i bgp$asic /usr/bin/$1
+            if [[ -n "$TSA_STATE_UPDATE" ]]; then 
+                sonic-cfggen -a "$TSA_STATE_UPDATE" -w -n $NAMESPACE_PREFIX$asic
+            else
+                docker exec -i bgp$asic /usr/bin/$1
```
[the remark this answers is not in the document]
Comment (tjchadaga) [reply dockers/docker-fpm-frr/base_image_files/TS:32]: this is now only being done for TSC. I will add a comment here

File: dockers/docker-fpm-frr/base_image_files/TSA [hunk 2]
```diff
@@ -1,12 +1,20 @@
...
+  # toggle the mux to standby if dualtor and any mux active
+  if
+  [[ "$(sonic-cfggen -d -v DEVICE_METADATA.localhost.subtype | tr [:upper:] [:lower:])" == *"dualtor"* ]] &&
+  [[ $(show mux status | grep active | wc -l) > 0 ]];
+  then
+    logger -t TSA -p user.info "Toggle all mux mode to standby"
+    sudo config mux mode standby all
+  fi
+
+  /usr/bin/TS TSA
```
[the remark this answers is not in the document]
Comment (tjchadaga) [reply dockers/docker-fpm-frr/base_image_files/TSA:17]: done

File: dockers/docker-fpm-frr/base_image_files/TSB [hunk 3]
```diff
@@ -1,10 +1,17 @@
...
+else
+  # toggle the mux to auto if dualtor
+  if [[ "$(sonic-cfggen -d -v DEVICE_METADATA.localhost.subtype | tr [:upper:] [:lower:])" == *"dualtor"* ]];
+  then
+    logger -t TSB -p user.info "Toggle all mux mode to auto"
+    sudo config mux mode auto all
+  fi
 
-/usr/bin/TS TSB
+  /usr/bin/TS TSB
```
[the remark this answers is not in the document]
Comment (tjchadaga) [reply dockers/docker-fpm-frr/base_image_files/TSB:14]: done

File: files/build_templates/init_cfg.json.j2 [hunk 4]
```diff
@@ -27,6 +27,11 @@
             "POLL_INTERVAL": "10000"
         }
     },
+    "BGP_DEVICE_GLOBAL": {
+        "STATE": {
+	   "tsa_enabled": "false"
```
Comment (qiluo-msft) [inline files/build_templates/init_cfg.json.j2:32]: >	    [](http://example.com/codeflow?start=0&length=4)

Please reformat json indentation. Do not mix spaces with tabs.

Comment (tjchadaga) [reply files/build_templates/init_cfg.json.j2]: thanks for the pointer, fixed the issue

File: src/sonic-bgpcfgd/bgpcfgd/managers_device_global.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            log_notice("DeviceGlobalCfgMgr:: Device isolated. Executing TSA")
+            cmd += self.get_tsa_routemaps(self.cfg_mgr.get_text())
+        else:
+            log_notice("DeviceGlobalCfgMgr:: Device un-isolated. Executing TSB")
+            cmd += self.get_tsb_routemaps(self.cfg_mgr.get_text())
+
+        self.cfg_mgr.push(cmd)
+        log_debug("DeviceGlobalCfgMgr::Done")
+
+    def get_tsa_routemaps(self, cmds):
```
Comment (StormLiangMS) [inline src/sonic-bgpcfgd/bgpcfgd/managers_device_global.py:70]: get_tsa_routemaps is very similar with get_tsb_routemaps, could we combine these two function with a variable for different template?
something like 
get_ts_routemaps(self, cmds, template):

File: src/sonic-bgpcfgd/bgpcfgd/managers_device_global.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        return self.__generate_routemaps_from_template(route_map_names, self.tsa_template)
+
+    def get_tsb_routemaps(self, cmds):
+        if not cmds:
+            return ""
+
+        route_map_names = self.__extract_out_route_map_names(cmds)
+        return self.__generate_routemaps_from_template(route_map_names, self.tsb_template)
+
+    def __generate_routemaps_from_template(self, route_map_names, template):
```
Comment (StormLiangMS) [inline src/sonic-bgpcfgd/bgpcfgd/managers_device_global.py:84]: The function of this mgr is very similar with TSA and TSB under the sonic-buildimage/dockers/docker-fpm-frr/, should we deprecate them?

File: src/sonic-bgpcfgd/bgpcfgd/managers_device_global.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+    def set_handler(self, key, data):
+        log_debug("DeviceGlobalCfgMgr:: set handler")
+        """ Handle device tsa_enabled state change """
+        if not data:
+            log_err("DeviceGlobalCfgMgr:: data is None")
+            return False
+
+        if "tsa_enabled" in data:
+            self.cfg_mgr.commit()
```
Comment (StormLiangMS) [inline src/sonic-bgpcfgd/bgpcfgd/managers_device_global.py:35]: What's the purpose of commit/update here?

File: src/sonic-bgpcfgd/bgpcfgd/managers_bgp.py [hunk 8]
```diff
@@ -23,6 +24,7 @@ def __init__(self, common_objs, base_template):
         tf = common_objs['tf']
         self.policy_template = tf.from_file(base_template + "policies.conf.j2")
         self.peergroup_template = tf.from_file(base_template + "peer-group.conf.j2")
+        self.device_global_cfgmgr = DeviceGlobalCfgMgr(common_objs, "CONFIG_DB", "BGP_DEVICE_GLOBAL")
```
Comment (StormLiangMS) [inline src/sonic-bgpcfgd/bgpcfgd/managers_bgp.py:27]: same here.

File: src/sonic-bgpcfgd/bgpcfgd/managers_device_global.py (hunk 7 above)
Comment (tjchadaga) [reply src/sonic-bgpcfgd/bgpcfgd/managers_device_global.py:35]: This is mainly for the reboot sequence where tsa_enabled is set after all peer-group configuration, but before it is actually committed to configDB. Commit is used here to ensure TSA routemap generation is done with all the pg config

File: src/sonic-bgpcfgd/bgpcfgd/managers_device_global.py (hunk 6 above)
Comment (tjchadaga) [reply src/sonic-bgpcfgd/bgpcfgd/managers_device_global.py:84]: Yes, TSA/B script under sonic-buildimage/dockers/docker-fpm-frr will no longer be invoked and can be deprecated after this change

**Review by StormLiangMS — APPROVED** [review]
LGTM

**Comment by qiluo-msft** [comment]
This commit could not be cleanly cherry-picked to 202012. Please submit another PR.

**Pull request merged**